### PR TITLE
add: verify signature on new state

### LINKED ISSUE #28: follower tick: check signature of NewState
in the follower tick, we should evaluate the signature of `NewState` for extra security

(even though we already check authentication on POST /validator-messages)

This actually goes deeper: you can enter arbitrary messages of any kind (`NewState` too) if you compromise the auth, which implies more attack options; for example, submitting an `ApproveState{ health: 'HEALTHY'  ...}` from the leader to the follower, to trick the market into thinking the channel is healthy

### PR DESCRIPTION
**Pull request by samparsky**
Fixes #28

### CODE REVIEW

File: adapters/dummy/index.js [hunk 1]
```diff
@@ -38,4 +38,18 @@ function getAuthFor(validator) {
 	else return Promise.reject(`no auth token for this identity: ${identity}`)
 }
 
-module.exports = { init, unlock, whoami, sign, getBalanceLeaf, sessionFromToken, getAuthFor, MerkleTree }
+function verify(stateRoot, signature) {
+	return signature === sign(stateRoot)
```
[the remark this answers is not in the document]
Comment (samparsky) [reply adapters/dummy/index.js]: I have changed the verify mechanism

File: services/validatorWorker/follower.js [hunk 2]
```diff
@@ -38,22 +38,28 @@ function onNewState(adapter, {channel, balances, newMsg, approveMsg}) {
...
-	return adapter.sign(stateRootRaw)
-	.then(function(signature) {
-		return persistAndPropagate(adapter, otherValidators, channel, {
-			type: 'ApproveState',
-			stateRoot: stateRoot,
-			isHealthy: isHealthy(balances, newBalances),
-			signature,
+	const { stateRoot, signature } = newMsg
+	// verify the signature of newMsg
+	return adapter.verify(stateRoot, signature)
```
Comment (Ivshti) [inline services/validatorWorker/follower.js]: inconsistent style: we don't use an additional ident for `.then` after promises, it's even that way in that code

File: services/validatorWorker/follower.js [hunk 3]
```diff
@@ -38,12 +38,17 @@ function onNewState(adapter, {channel, balances, newMsg, approveMsg}) {
...
-	// currently, in order to POST to /validator-messages, we require that you're a validator
-	// but this means that if a validator's auth token leaks, that can be used to trick followers into signing any new state that is a valid state transition
 	const whoami = adapter.whoami()
 	const otherValidators = channel.spec.validators.filter(v => v.id != whoami)
-	const stateRoot = newMsg.stateRoot
+	const {stateRoot, signature } = newMsg
+
+	// verify the signature of newMsg
+	assert.equal(
+		adapter.verify(stateRoot, signature),
```
[the remark this answers is not in the document]
Comment (samparsky) [reply services/validatorWorker/follower.js]: Yeah, changed it to return promise value, verifyMessage returns a string( address ) its not promised according to its doc

File: services/validatorWorker/follower.js [hunk 4]
```diff
@@ -38,22 +38,28 @@ function onNewState(adapter, {channel, balances, newMsg, approveMsg}) {
...
-			type: 'ApproveState',
-			stateRoot: stateRoot,
-			isHealthy: isHealthy(balances, newBalances),
-			signature,
+	const { stateRoot, signature } = newMsg
+	// verify the signature of newMsg
+	return adapter.verify(stateRoot, signature)
+		.then(function(res){
+			if(!res) {
+				return Promise.reject("Invalid signature on NEWSTATE")
```
Comment (Ivshti) [inline services/validatorWorker/follower.js]: this error should actually be handled in the same way as this:
https://github.com/AdExNetwork/adex-validator-stack-js/blob/fa4c49445dcbe64f2d6ecc623103eb1414b03c77/services/validatorWorker/follower.js#L36

In other words, the tick executed successfully, but it detected a misbehavior and therefore `nothingNew` is true

File: services/validatorWorker/follower.js (hunk 2 above)
Comment (samparsky) [reply services/validatorWorker/follower.js]: okay will remove the identation

File: services/validatorWorker/follower.js (hunk 4 above)
Comment (samparsky) [reply services/validatorWorker/follower.js]: okay, made the change

**Pull request merged**